Indicators/TimeSeriesIndicator.cs

```diff
@@ -0,0 +1,158 @@
+/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// The base class for any Time Series-type indicator, containing methods common to most of such models.
+    /// </summary>
+    public abstract class TimeSeriesIndicator : IndicatorBase<IndicatorDataPoint>, IIndicatorWarmUpPeriodProvider
+    {
+        private double[] _diffHeads;
+
+        /// <summary>
+        /// A constructor for a basic Time Series indicator.
+        /// </summary>
+        /// <param name="name">The name of this indicator</param>
+        protected TimeSeriesIndicator(string name)
+            : base(name)
+        {
+        }
+
+        /// <summary>
+        /// Gets a flag indicating when this indicator is ready and fully initialized
+        /// </summary>
+        public override bool IsReady { get; }
+
+        /// <summary>
+        /// Required period, in data points, for the indicator to be ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod { get; }
```

Comment:
Should be able to directly remove `public override bool IsReady { get; }`
`WarmUpPeriod ` should be abstract so derivative can override. Using `new` (at `ArimaIndicator.WarmUpPeriod`) is usually a code smell (something is wrong 😄)
nit - let's move properties above constructor